### PR TITLE
[MISC] use new CPO-way to define seqan3::enumeration_names

### PR DESCRIPTION
**Pull request by marehr**
Part of https://github.com/seqan/product_backlog/issues/298.

### CODE REVIEW

File: include/seqan3/argument_parser/auxiliary.hpp [hunk 1]
```diff
@@ -73,38 +73,49 @@ namespace seqan3::detail::adl_only
...
+struct enumeration_names_cpo : public detail::customisation_point_object<enumeration_names_cpo<option_t>, 1>
 {
-    //!\brief `option_t` with cvref removed and possibly wrapped in std::type_identity.
-    using s_option_t = std::conditional_t<std::is_nothrow_default_constructible_v<std::remove_cvref_t<option_t>> &&
-                                          seqan3::is_constexpr_default_constructible_v<std::remove_cvref_t<option_t>>,
-                                          std::remove_cvref_t<option_t>,
-                                          std::type_identity<option_t>>;
+    //!\brief CRTP base class seqan3::detail::customisation_point_object.
+    using base_t = detail::customisation_point_object<enumeration_names_cpo<option_t>, 1>;
+    //!\brief Only this class is allowed to import the constructors from #base_t. (CRTP safety idiom)
```
Comment (MitraDarja) [inline include/seqan3/argument_parser/auxiliary.hpp:83]: What is #base_t?

File: include/seqan3/argument_parser/auxiliary.hpp [hunk 2]
```diff
@@ -73,38 +73,49 @@ namespace seqan3::detail::adl_only
...
-    SEQAN3_CPO_IMPL(0, (enumeration_names(v))) // ADL
+    /*!\brief If `alphabet_type` isn't std::is_nothrow_default_constructible, enumeration_names will be called with
+     *        std::type_identity instead of a default constructed alphabet.
+     */
+    template <typename option_type>
+    using option_or_type_identity
+        = std::conditional_t<std::is_nothrow_default_constructible_v<std::remove_cvref_t<option_type>> &&
+                             seqan3::is_constexpr_default_constructible_v<std::remove_cvref_t<option_type>>,
+                             std::remove_cvref_t<option_type>,
+                             std::type_identity<option_type>>;
```
Comment (MitraDarja) [inline include/seqan3/argument_parser/auxiliary.hpp:94]: Shouldn't this be aligned to std::remove_cvref_t<option_type>?

File: include/seqan3/argument_parser/auxiliary.hpp [hunk 3]
```diff
@@ -73,38 +73,49 @@ namespace seqan3::detail::adl_only
...
-            std::same_as<decltype(impl(priority_tag<1>{}, s_option_t{}, dummy{})),
-                         std::unordered_map<std::string_view, std::remove_cvref_t<option_t>>>;
-        }
-    //!\endcond
-    auto operator()() const
-    {
-        return impl(priority_tag<1>{}, s_option_t{});
-    }
+    /*!\brief CPO overload (1. out of 2 checks): explicit customisation via `seqan3::custom::argument_parsing`
+     * \tparam option_type The type of the option. (Needed to defer instantiation for incomplete types)
```
Comment (MitraDarja) [inline include/seqan3/argument_parser/auxiliary.hpp:97]: Dot after types?

File: include/seqan3/argument_parser/auxiliary.hpp [hunk 4]
```diff
@@ -73,38 +73,49 @@ namespace seqan3::detail::adl_only
...
+     */
+    template <typename option_type = option_t>
+    static constexpr auto SEQAN3_CPO_OVERLOAD(priority_tag<1>)
+    (
+        /*return*/ seqan3::custom::argument_parsing<option_type>::enumeration_names /*;*/
```
Comment (MitraDarja) [inline include/seqan3/argument_parser/auxiliary.hpp:102]: What are these comments for?

File: include/seqan3/argument_parser/auxiliary.hpp [hunk 5]
```diff
@@ -73,38 +73,49 @@ namespace seqan3::detail::adl_only
...
+     *
+     * \details
+     *
+     * If the option_type isn't std::is_nothrow_default_constructible,
+     * `enumeration_names(std::type_identity<alphabet_type>{})` will be called.
+     */
+    template <typename option_type = option_t>
+    static constexpr auto SEQAN3_CPO_OVERLOAD(priority_tag<0>)
+    (
+        /*return*/ enumeration_names(option_or_type_identity<option_type>{}) /*;*/
```
Comment (MitraDarja) [inline include/seqan3/argument_parser/auxiliary.hpp:117]: Here, as well, why do we need these comments?

**Review by MitraDarja — CHANGES_REQUESTED** [review]
LGTM, but some minor things.

File: include/seqan3/argument_parser/auxiliary.hpp (hunk 1 above)
Comment (marehr) [reply include/seqan3/argument_parser/auxiliary.hpp:83]: `#base_t` is a doxygen command to force a link to base_t. Unfortunately, `using base_t::base_t;` isn't shown in the documentation.

File: include/seqan3/argument_parser/auxiliary.hpp (hunk 4 above)
Comment (marehr) [reply include/seqan3/argument_parser/auxiliary.hpp:102]: As you can see this isn't a regular function it is in fact a macro, I hoped that the `/**/` comment help to understand that the macro will resolve to a normal function that return that expression.

File: include/seqan3/argument_parser/auxiliary.hpp (hunk 3 above)
Comment (marehr) [reply include/seqan3/argument_parser/auxiliary.hpp:97]: I added a todo to change this afterwards https://github.com/seqan/product_backlog/issues/298#issuecomment-805964030 (as I have a lot of PRs already merged / open that are affected by this)

I'll change this after everything was merged. Good catch!

File: include/seqan3/argument_parser/auxiliary.hpp (hunk 2 above)
Comment (marehr) [reply include/seqan3/argument_parser/auxiliary.hpp:94]: I'm sorry I'm not sure which `std::remove_cvref_t<option_type>` you mean, currently it is aligned like this:

```c++
template <typename option_type>
using option_or_type_identity
    = std::conditional_t<long_condition1 &&
                         long_condition2,
                         type1,
                         type2>>;
```

**Review by MitraDarja — APPROVED** [review]
Thanks for the explainations. :)

**Review by eseiler — APPROVED** [review]

**Pull request merged**